### PR TITLE
ARROW-16652: [Python] Cast compute kernel segfaults when called with a Table

### PR DESCRIPTION
**Pull request by kshitij12345**
Repro Script:
```python
import pyarrow as pa
import pyarrow.compute as pc

table = pa.table({'a': [1, 2]})

pc.cast(table, pa.int64())
```

Post PR we get,
```
pyarrow.lib.ArrowInvalid: Tried executing function with non-value type: Table
```

which is the same error we get if we replace `pc.cast(table, pa.int64())` with `pc.abs(table)`.

### CODE REVIEW

File: cpp/src/arrow/compute/cast.cc
Comment (raulcd) [inline cpp/src/arrow/compute/cast.cc:99]: Should we ensure the argument is always an Array? Something like:
```
  switch (args[0].kind()) {
    case Datum::ARRAY:
      // Implementation
      break;
    default:
      return Status::NotImplemented("Cast arguments should be array-like");
```

**Review by raulcd — COMMENTED** [review]
Thanks for the PR @kshitij12345 !
Maybe we can also add a test on the python test_compute.py module that ensure the exception is raised to validate the use case.

File: cpp/src/arrow/compute/cast.cc
Comment (kshitij12345) [reply cpp/src/arrow/compute/cast.cc:99]: I think it should be done to all functions which don't support this. Probably worth an issue by itself?

**Comment by kshitij12345** [comment]
> Maybe we can also add a test on the python test_compute.py module that ensure the exception is raised to validate the use case.

Right. Just wanted to get a review if the approach made sense. Will add a test soon. Thanks!

**Comment by kshitij12345** [comment]
cc: @AlenkaF

**Comment by kshitij12345** [comment]
Gentle ping @raulcd

**Comment by raulcd** [comment]
Thanks @kshitij12345 ! Let me ping someone more familiar with this part of the codebase to follow it up, @westonpace maybe?

**Comment by kshitij12345** [comment]
Gentle ping :)

**Review by westonpace — APPROVED** [review]
I think this is a good fix.  I believe cast functions support chunked array and scalars too (so I might be wrong) so I don't think we should assert that it is an array.  It also looks like the cast functions themselves validate their inputs (that is where the new error message is coming from) so I'm not sure there is a larger issue to address.  It still might make sense to check if it is a record batch, table, or none and just flat-out reject in that case with an even clearer message "cast is only supported on a single column" though I'm happy to approve the PR as-is.

**Comment by AlenkaF** [comment]
Thank you @westonpace for the review!

@kshitij12345 do you want to add a check for a record batch or table as Weston suggested or you want to leave it as is?
I would maybe try rebasing to see if the CI behaves any better.

**Comment by kshitij12345** [comment]
@AlenkaF @westonpace 

Thanks for the suggestion. I think I would like to look at it separately as I have less bandwidth this month.

**Comment by AlenkaF** [comment]
The CI failures are not related ([R dcukdb has a JIRA issue opened ](https://issues.apache.org/jira/browse/ARROW-17643)and LLVM can be seen on master also). So I will go ahead and merge the PR after getting the approval from @raulcd also.

Thank you @kshitij12345 for working on this and all other issues!

**Review by raulcd — APPROVED** [review]
Thanks for the PR!

**Pull request merged**
